freva-rest/src/freva_rest/auth.py

```diff
@@ -1,175 +1,93 @@
-        response = await client.post(
-            server_config.keycloak_overview["token_endpoint"], data=data
-        )
-        if response.status != 200:
-            raise HTTPException(
-                status_code=response.status,
-                detail="Invalid username or password",
-            )
-        return Token(**(await response.json()))
+@app.get("/api/auth/v2/status", tags=["Authentication"])
+async def get_token_status(
+    id_token: IDToken = Security(auth.required),
+) -> TokenPayload:
+    """Check the status of an access token."""
+    return cast(TokenPayload, id_token)
 
 
-oauth2_scheme = OAuth2AuthorizationCodeBearer(
-    authorizationUrl=server_config.keycloak_overview["authorization_endpoint"],
-    tokenUrl=server_config.keycloak_overview["token_endpoint"],
+@app.get(
+    "/api/auth/v2/.well-known/openid-configuration",
+    tags=["Authentication"],
+    response_class=RedirectResponse,
 )
+async def open_id_config() -> RedirectResponse:
+    """Get the open id connect configuration."""
+    return RedirectResponse(server_config.oidc_discovery_url)
 
 
-class TokenPayload(BaseModel):
-    """Model representing the payload of a JWT token."""
-
-    sub: str
-    exp: int
-    email: Optional[str] = None
-
-
-async def get_current_user(
-    token: str = Depends(oauth2_scheme),
-    verify_exp: bool = True,
-) -> TokenPayload:
-    """Get the current user by verifying the JWT token.
-
-    Parameters
-    ----------
-    token (str):
-        The JWT token from the Authorization header.
-
-    Returns
-    -------
-    TokenPayload: The payload data from the token.
-
-    Raises
-    ------
-    HTTPException: If the token is invalid or verification fails.
-    """
-    credentials_exception = HTTPException(
-        status_code=status.HTTP_401_UNAUTHORIZED,
-        detail="Could not validate credentials",
-        headers={"WWW-Authenticate": "Bearer"},
-    )
-
-    try:
-        jwks_client = jwt.PyJWKClient(
-            server_config.keycloak_overview["jwks_uri"],
-            headers={"User-agent": "custom-user-agent"},
-        )
-        signing_key = jwks_client.get_signing_key_from_jwt(token)
-        payload = jwt.decode(
-            token,
-            signing_key.key,
-            algorithms=["RS256"],
-            options={"verify_exp": verify_exp, "verify_aud": False},
-        )
-        token_data = TokenPayload(**payload)
-    except jwt.exceptions.PyJWTError as error:
-        logger.critical(error)
-        raise credentials_exception from None
-
-    return token_data
-
-
-async def get_user_status(token: str = Depends(oauth2_scheme)) -> TokenPayload:
-    """Get the status of an existing user token."""
-    return await get_current_user(token, verify_exp=False)
-
-
-@app.post(
-    "/api/auth/v2/refresh", response_model=Token, tags=["Authentication"]
-)
-async def refresh_access_token(
+@app.post("/api/auth/v2/token", tags=["Authentication"])
+async def fetch_or_refresh_token(
```

Comment:
instead of parsing the request before fetch_or_refresh_token stage, if we get the whole request as Request here, we don't need to define and take care of so many attrs complicated typing i think:

```python
from fastapi import Request
async def fetch_or_refresh_token(request: Request):
    data = {
        "client_id": request.client_id or server_config.oidc_client,
        "client_secret": request.client_secret or os.getenv("OIDC_CLIENT_SECRET", ""),
        "grant_type": request.grant_type,
        "username": request.username,
        "password": request.password,
        "refresh_token": request.refresh_token,
    }
    data = {k: v for k, v in data.items() if v}
```

tbh I haven't tried in fastapi but it seems it has to be feasible ...

Reply:
Yes, that was indeed my first approach. But then I decided to follow this approach because it integrates much better with the swagger help.